### PR TITLE
deployer for wordpress

### PR DESCRIPTION
**Pull request by wuyuxin511**
the yaml file read in process was blocked and has something more to do.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -23,6 +23,8 @@ repositories {
 dependencies {
     // This dependency is used by the application.
     implementation 'com.google.guava:guava:29.0-jre'
+    compile 'io.kubernetes:client-java:8.0.2'
```
Comment (magl0) [inline build.gradle]: use `implementation` here so that it is consistent across the file.

File: application/wordpress/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+secretGenerator:
```
Comment (magl0) [inline application/wordpress/kustomization.yaml]: This file is not needed anymore right?

File: application/wordpress/mysql-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      labels:
+        app: wordpress
+        tier: mysql
+    spec:
+      containers:
+      - image: mysql:5.6
+        name: mysql
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          value: password
```
Comment (magl0) [inline application/wordpress/mysql-deployment.yaml:52]: Ideally in the future we want to make this a template parameter as well.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 4]
```diff
@@ -16,12 +16,180 @@
...
+import io.kubernetes.client.util.KubeConfig;
+import io.kubernetes.client.util.Yaml;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 public class App {
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java:34]: Please follow google java style, e.g. set indentation to 2 space, etc. You can download the `google-java-format` plugin and reformat the entire project using that tool.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 5]
```diff
@@ -16,12 +16,180 @@
...
+        System.out.println(Yaml.dump(svc));
+        */
+
+
+        // file path to KubeConfig
+        String kubeConfigPath = System.getProperty("user.home") + "/.kube/config";
+
+        // loading the out-of-cluster config, a kubeconfig from file-system
+        ApiClient client =
+                ClientBuilder.kubeconfig(KubeConfig.loadKubeConfig(new FileReader(kubeConfigPath))).build();
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: I think you can simply use `ApiClient client = Config.defaultClient();` to get an instance of `ApiClient`. `defaultClient` will work both locally and within a cluster. See comment [here](https://github.com/kubernetes-client/java/blob/cbc1d385dfa02da38e07e4b22498971b7b464680/util/src/main/java/io/kubernetes/client/util/Config.java#L103-L108).

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 6]
```diff
@@ -16,12 +16,180 @@
...
+                        .withPort(80)
+                        .endPort()
+                        .withSelector(m2)
+                        .addToSelector("app","wordpress")
+                        .addToSelector("tier", "frontend")
+                        .withType("LoadBalancer")
+                        .endSpec()
+                        .build();
+        System.out.println(Yaml.dump(svc));
+        */
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: Is this for future reference? If so, you can replace this with a comment that links to the api example. This helps to make the source code clean.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 7]
```diff
@@ -16,12 +16,180 @@
...
+                .build();
+        helloCmd.parse(args);
+
+        String appName = jArgs.getName();
+        String appVersion = jArgs.getVersion();
+
+        System.out.println("Hello " + appName + " " + appVersion);
+
+        /* Logic to get the relative application config */
+        String configPath = System.getProperty("user.dir") + "/application/" + appName + "/";
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: Doesn't necessarily block this PR, but as we mentioned in the design slides, you can replace this with a command line flag.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 8]
```diff
@@ -16,12 +16,180 @@
...
+                File[] files = file.listFiles();
+                for (File f : files) {
+                    if (!f.isDirectory())
+                        /* TODO: parse all config files under a certain app name to Kubernetes Object here, maybe use index to represent each config file*/
+                        System.out.println(f);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: You can simplify this code with `MoreFiles.fileTraverser()` from Guava. Guava is the common library released by Google with a lot of useful utilities that can make your life easier. See the document for this util [here](https://guava.dev/releases/29.0-jre/api/docs/com/google/common/io/MoreFiles.html#fileTraverser--).

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 9]
```diff
@@ -16,12 +16,180 @@
...
+        // TODO: load kustomization or secret file (unfixed problem in creating namespaced secret object)
+        /*
+        String secretPath = System.getProperty("user.dir") + "/application/wordpress/mysql-pass.yaml";
+        File secretFile = new File(secretPath);
+        V1Secret secret = (V1Secret) Yaml.load(secretFile);
+        System.out.println(secret);
+        */
+
+        /* Initialize apis */
+        CoreV1Api api = new CoreV1Api();
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: use a more descriptive name like `coreV1Api`.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 10]
```diff
@@ -16,12 +16,180 @@
...
+        AppsV1Api api1 = new AppsV1Api();
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: ditto

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 11]
```diff
@@ -16,12 +16,180 @@
...
+        V1Service createResultSvc = api.createNamespacedService("default", mysqlSvc, null, null, null);
+        V1Service createResultWordpressSvc = api.createNamespacedService("default", wordpressSVC, null, null, null);
+
+        /* Create persistent volume claims */
+        V1PersistentVolumeClaim createResultPvc= api.createNamespacedPersistentVolumeClaim("default", mysqlPvc, null,null,null);
+        V1PersistentVolumeClaim createResultWordpressPvc= api.createNamespacedPersistentVolumeClaim("default", wordpressPVC, null,null,null);
+
+        /* Deploy */
+        V1Deployment createResultDeployment= api1.createNamespacedDeployment("default", mysqlDeployment, null,null,null);
+        V1Deployment createResultWordpressDeployment= api1.createNamespacedDeployment("default", wordpressDeployment, null,null,null);/*/
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: Remove the `/*/` at the end of this line.

Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: You can simplify this code a lot by using a map from `Object.getClass()` to the API call. You can find a similar solution in [this github issue](https://github.com/kubernetes-client/java/issues/965).

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 12]
```diff
@@ -16,12 +16,180 @@
...
+
+        /* Load mysql config file */
+        String mysqlConfigPath = System.getProperty("user.dir") + "/application/wordpress/mysql-deployment.yaml";
+        File mysqlFile = new File(mysqlConfigPath);
+        List<Object> objectList4Mysql = (List<Object>) Yaml.loadAll(mysqlFile);
+
+        /* Load wordpress config file */
+        String wordpressConfigPath = System.getProperty("user.dir") + "/application/wordpress/wordpress-deployment.yaml";
+        File wordpressConfig = new File(wordpressConfigPath);
+        List<Object> objectList4Wordpress = (List<Object>) Yaml.loadAll(wordpressConfig);
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: There is a comment on line 61 that you will load the config while traversing the file tree. So maybe move this block of code up to make it cleaner.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 13]
```diff
@@ -16,12 +16,180 @@
...
+        */
+
+        /* Initialize apis */
+        CoreV1Api api = new CoreV1Api();
+        AppsV1Api api1 = new AppsV1Api();
+
+        /* Load mysql config file */
+        String mysqlConfigPath = System.getProperty("user.dir") + "/application/wordpress/mysql-deployment.yaml";
+        File mysqlFile = new File(mysqlConfigPath);
+        List<Object> objectList4Mysql = (List<Object>) Yaml.loadAll(mysqlFile);
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: Prefer to use immutable collections from Guava. In this case you can replace the code with:

```java
ImmutableList<Object> objectList4Mysql = ImmutableList.copyOf((List<Object>) Yaml.loadAll(mysqlFile));
```

This [page](https://immutables.github.io/immutable.html) have a great summary on the benefits of immutable objects.

File: src/main/java/tsunami/security/scanner/utilities/App.java [hunk 14]
```diff
@@ -16,12 +16,180 @@
...
+                        curSvc.getMetadata().getName(),
+                        "default",
+                        null,
+                        null,
+                        null,
+                        null,
+                        null,
+                        new V1DeleteOptions());
+        System.out.println(deleteResult);
+        */
```
Comment (magl0) [inline src/main/java/tsunami/security/scanner/utilities/App.java]: Remove this as well.

**Review by magl0 — COMMENTED** [review]
Great job for your first Java code PR!

File: src/main/java/tsunami/security/scanner/utilities/App.java (hunk 4 above)
Comment (magl0) [reply src/main/java/tsunami/security/scanner/utilities/App.java:34]: I mean the `google-java-format` plugin in intellij.

Comment (magl0) [reply src/main/java/tsunami/security/scanner/utilities/App.java:34]: And Google's Java style is at https://google.github.io/styleguide/javaguide.html.

File: application/wordpress/kustomization.yaml (hunk 2 above)
Comment (wuyuxin511) [reply application/wordpress/kustomization.yaml]: yeah, already removed

File: src/main/java/tsunami/security/scanner/utilities/App.java (hunk 5 above)
Comment (wuyuxin511) [reply src/main/java/tsunami/security/scanner/utilities/App.java]: done

File: src/main/java/tsunami/security/scanner/utilities/App.java (hunk 11 above)
Comment (magl0) [reply src/main/java/tsunami/security/scanner/utilities/App.java]: Let's resolve this in a follow up PR.

**Comment by magl0** [comment]
Thanks for your work! LGTM

**Pull request merged**